File: webapp/src/mifare/card-layout.ts
/**
 * Maps one NFAR chunk onto the usable data blocks of a Mifare Classic 1K card.
 * Layout is raw NFAR-native: serialized chunk bytes are written sequentially
 * across the 47 usable blocks (block 0 and every sector trailer skipped),
 * zero-padding the final block. The NFAR header is self-delimiting, so no
 * per-block framing is added.
 */

import { NFAR_MAGIC, NFAR_VERSION, TOTAL_OVERHEAD, NfarFormatError } from '../chunk.js';

export const BLOCK_SIZE = 16;

export const USABLE_BLOCK_INDEXES: readonly number[] = Object.freeze(
  Array.from({ length: 64 }, (_, b) => b).filter((b) => b !== 0 && b % 4 !== 3),
);

export const CARD_CAPACITY_BYTES = USABLE_BLOCK_INDEXES.length * BLOCK_SIZE; // 752
export const CARD_PAYLOAD_SIZE = CARD_CAPACITY_BYTES - TOTAL_OVERHEAD; // 720

export class CardCapacityError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CardCapacityError';
  }
}

export function chunkToBlocks(chunkBytes: Uint8Array): { block: number; data: Uint8Array }[] {
  if (chunkBytes.length > CARD_CAPACITY_BYTES) {
    throw new CardCapacityError(
      `Chunk is ${chunkBytes.length} bytes; a Mifare Classic 1K card holds ${CARD_CAPACITY_BYTES}`,
    );
  }
  const blockCount = Math.ceil(chunkBytes.length / BLOCK_SIZE);
  const out: { block: number; data: Uint8Array }[] = [];
  for (let i = 0; i < blockCount; i++) {
    const data = new Uint8Array(BLOCK_SIZE); // zero-filled -> pads the last block
    data.set(chunkBytes.subarray(i * BLOCK_SIZE, (i + 1) * BLOCK_SIZE));
    out.push({ block: USABLE_BLOCK_INDEXES[i]!, data });
  }
  return out;
}

export function firstBlockIsNfar(block1: Uint8Array): boolean {
  if (block1.length < NFAR_MAGIC.length + 1) return false;
  for (let i = 0; i < NFAR_MAGIC.length; i++) {
    if (block1[i] !== NFAR_MAGIC[i]) return false;
  }
  return block1[NFAR_MAGIC.length] === NFAR_VERSION;
}

export function nfarTotalLength(header: Uint8Array): number {
  if (!firstBlockIsNfar(header)) {
    throw new NfarFormatError('Not an NFAR card: magic or version mismatch');
  }
  if (header.length < 28) {
    throw new NfarFormatError(`Header too short: need 28 bytes, got ${header.length}`);
  }
  const view = new DataView(header.buffer, header.byteOffset, header.byteLength);
  const payloadSize = view.getUint16(26); // big-endian, per NFAR header layout
  return TOTAL_OVERHEAD + payloadSize;
}

export function assembleChunkFromBlocks(orderedBlockData: Uint8Array[], totalLength: number): Uint8Array {
  const flat = new Uint8Array(orderedBlockData.length * BLOCK_SIZE);
  orderedBlockData.forEach((b, i) => flat.set(b, i * BLOCK_SIZE));
  return flat.slice(0, totalLength);
}

File: lib/core/mifare/card_layout.dart
import 'dart:typed_data';

import '../constants/nfar_format.dart';

/// Maps one NFAR chunk onto the usable data blocks of a Mifare Classic 1K card.
///
/// Deliberate port of `webapp/src/mifare/card-layout.ts` — same constants, same
/// ordering — so cards written by either app are byte-identical. Change both or
/// neither.
///
/// Layout is raw NFAR-native: serialized chunk bytes run sequentially across the
/// 47 usable blocks (block 0 and every sector trailer skipped), zero-padding the
/// final block. The NFAR header is self-delimiting, so no per-block framing is
/// added.
const int blockSize = 16;

/// Blocks 0-63 minus block 0 (manufacturer) and every `b % 4 == 3` (sector
/// trailer, holds the keys — writing one can brick the card).
final List<int> usableBlockIndexes = List.unmodifiable(
  List<int>.generate(64, (b) => b).where((b) => b != 0 && b % 4 != 3).toList(),
);

final int cardCapacityBytes = usableBlockIndexes.length * blockSize; // 752
final int cardPayloadSize = cardCapacityBytes - NfarHeaderSize.total; // 720

/// A chunk does not fit a Mifare Classic 1K card.
class MifareCapacityException implements Exception {
  MifareCapacityException(this.message);
  final String message;
  @override
  String toString() => 'MifareCapacityException: $message';
}

/// One block-sized write: which block, and the 16 bytes to put there.
class MifareBlockWrite {
  const MifareBlockWrite(this.block, this.data);
  final int block;
  final Uint8List data;
}

/// Split serialized chunk bytes across the usable blocks, zero-padding the last.
List<MifareBlockWrite> chunkToBlocks(Uint8List chunkBytes) {
  if (chunkBytes.length > cardCapacityBytes) {
    throw MifareCapacityException(
      'Chunk is ${chunkBytes.length} bytes; a Mifare Classic 1K card holds '
      '$cardCapacityBytes',
    );
  }
  final blockCount = (chunkBytes.length + blockSize - 1) ~/ blockSize;
  final out = <MifareBlockWrite>[];
  for (var i = 0; i < blockCount; i++) {
    final data = Uint8List(blockSize); // zero-filled -> pads the last block
    final start = i * blockSize;
    final end = (start + blockSize).clamp(0, chunkBytes.length);
    data.setAll(0, chunkBytes.sublist(start, end));
    out.add(MifareBlockWrite(usableBlockIndexes[i], data));
  }
  return out;
}

/// Whether the first usable block starts an NFAR chunk.
bool firstBlockIsNfar(Uint8List block1) {
  if (block1.length < nfarMagic.length + 1) return false;
  for (var i = 0; i < nfarMagic.length; i++) {
    if (block1[i] != nfarMagic[i]) return false;
  }
  return block1[nfarMagic.length] == nfarVersion;
}

/// Total serialized length declared by an NFAR header.
int nfarTotalLength(Uint8List header) {
  if (!firstBlockIsNfar(header)) {
    throw const FormatException('Not an NFAR card: magic or version mismatch');
  }
  if (header.length < NfarHeaderSize.total) {
    throw FormatException(
      'Header too short: need ${NfarHeaderSize.total} bytes, '
      'got ${header.length}',
    );
  }
  final payloadSize = ByteData.sublistView(header).getUint16(26); // big-endian
  return NfarHeaderSize.total + payloadSize;
}
